### ISSUE TITLE
please support Intel MKL, IPP and DAAL since they are freely available now

### ISSUE THREAD

**Issue by jeffhammond** [travis-ci/travis-ci#5381]
It would be nice if Travis could support MKL, IPP and DAAL since all of these (and TBB) are _freely available to all_ via [Community Licensing](https://software.intel.com/sites/campaigns/nest/).

I work for Intel.  Please let me know if their are logistical issues that need resolving here and I will bring them to the attention of the right people.

**Comment by BanzaiMan** [comment]
We offer paid services for private repositories, which, for the most part, is for commercial software. This, I believe, violates 4.G.ii. of the Community License.

**Comment by jeffhammond** [comment]
I don't think 4.G.ii applies:

"Community Licensing is provided as a Named-User License per our standard end user licensing agreement, without Intel Premier support, and with no fees, no royalties, and no restrictions on company or project sizes."

IANAL but I think the non commercial restrictions don't apply here. 

If Travis needs official clarification/confirmation from Intel, I'll try to take care of that.

**Comment by BanzaiMan** [comment]
@jeffhammond Sure thing. Please email us at support@travis-ci.com. We'll chat. :-)

**Comment by izaid** [comment]
Was there ever a resolution to this? I'm sure many people, myself included, would like easy access to Intel MKL.

**Comment by jeffhammond** [comment]
No, but only because of a lack of time on my part, not because of a lack of interest.  Sorry.

**Comment by izaid** [comment]
Anyway I can help? Or is this a question of you emailing them?

**Comment by jeffhammond** [comment]
Write me email (it's on my Github profile) and we can discuss.

**Comment by izaid** [comment]
Emailed.

**Comment by jeffhammond** [comment]
Sorry for the unreasonably long delay.  I have started making progress on this internally.  Will reach out once we are ready.

**Comment by Yangqing** [comment]
I am wondering if there are any updates for this :) We would love to be able to contbuild test against MKL as it gives us a sweet speedup on linear algebra. Background: at Facebook right now and am in the process of open-sourcing a bunch of stuff. @jeffhammond - is there something I can help with? Please let me know.

**Comment by jeffhammond** [comment]
Yes!  See https://github.com/travis-ci/travis-ci/issues/4604#issuecomment-270669847.  Give credit to @rscohn2 and coworkers, not me.

**Comment by jeffhammond** [comment]
@Yangqing And I recognize that what has been done is not the same as whitelisting, but it is still quite useful.

**Comment by rscohn2** [comment]
You can now download development packages for MKL, IPP with yum and apt. They include headers and static libraries.

Yum repo: https://software.intel.com/en-us/articles/installing-intel-free-libs-and-python-yum-repo
Apt repo: https://software.intel.com/en-us/articles/installing-intel-free-libs-and-python-apt-repo

We also simplified the licensing to be close to a MIT-style license. It was already OK to use MKL for commercial use with the free license, but now you don't need a lawyer to tell you that :).

My project is Intel Python, which is also available in yum/apt under the same license. It's based on Anaconda and includes optimization from MKL, DAAL, TBB to make scipy/numpy/sklearn make use of SIMD and multi-core.

It should be possible to install these packages with travis-ci if you are using VM's. It would be great if they could be used from containers, but since you have to add extra repos/keys, it is not a simple case of adding them to the whitelist. If @jeffhammond or someone else knows the best way to make this happen, please let me know.

**Comment by ismail** [comment]
@rscohn2 Note that apt-repos seems to be not working, at least for me only intelpython repo worked:

```
W: Invalid 'Date' entry in Release file /var/lib/apt/lists/partial/apt.repos.intel.com_intelpython_binary_InRelease
W: The repository 'http://apt.repos.intel.com/mkl stable Release' does not have a Release file.
N: Data from such a repository can't be authenticated and is therefore potentially dangerous to use.
N: See apt-secure(8) manpage for repository creation and user configuration details.
W: The repository 'http://apt.repos.intel.com/ipp stable Release' does not have a Release file.
N: Data from such a repository can't be authenticated and is therefore potentially dangerous to use.
N: See apt-secure(8) manpage for repository creation and user configuration details.
W: The repository 'http://apt.repos.intel.com/tbb stable Release' does not have a Release file.
N: Data from such a repository can't be authenticated and is therefore potentially dangerous to use.
N: See apt-secure(8) manpage for repository creation and user configuration details.
W: The repository 'http://apt.repos.intel.com/daal stable Release' does not have a Release file.
N: Data from such a repository can't be authenticated and is therefore potentially dangerous to use.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: Failed to fetch http://apt.repos.intel.com/mkl/dists/stable/main/binary-i386/Packages  404  Not Found [IP: 23.62.98.88 80]
E: Failed to fetch http://apt.repos.intel.com/ipp/dists/stable/main/binary-amd64/Packages  404  Not Found [IP: 23.62.98.88 80]
E: Failed to fetch http://apt.repos.intel.com/tbb/dists/stable/main/binary-amd64/Packages  404  Not Found [IP: 23.62.98.88 80]
E: Failed to fetch http://apt.repos.intel.com/daal/dists/stable/main/binary-i386/Packages  404  Not Found [IP: 23.62.98.88 80]
E: Some index files failed to download. They have been ignored, or old ones used instead.

```

**Comment by rscohn2** [comment]
@ismail 

I reproduced the problem in a docker container and reported to the maintainers of the repo. I will let you know when it is fixed.

**Comment by rscohn2** [comment]
@ismail there were some documentation and configuration issues. They have all been resolved. Please try again. Thanks.

Robert

**Comment by ismail** [comment]
@rscohn2 indeed it works, just a warning left to fix:

`W: Invalid 'Date' entry in Release file /var/lib/apt/lists/apt.repos.intel.com_intelpython_binary_InRelease`

Thanks a lot!

**Comment by rscohn2** [comment]
I opened an issue to add the intel repo's https://github.com/travis-ci/apt-source-whitelist/issues/367
After that is done I will submit issues to add the individual packages.

**Comment by faroit** [comment]
can someone point me to `.travis.yml` example where mkl is successfully deployed? I couldn't get tests to work that require pytorch and mkl. E.g. see [here](https://travis-ci.com/sigsep/open-unmix-pytorch/jobs/225161930)

**Comment by jeffhammond** [comment]
See https://travis-ci.org/nemequ/icc-travis/jobs/149894214 and related.  I can't figure out why @nemequ said it isn't working in the banner (https://github.com/nemequ/icc-travis/issues/14) but I will investigate.

On a related note, you might find https://software.intel.com/en-us/articles/installing-intel-free-libs-and-python-yum-repo or https://software.intel.com/en-us/articles/installing-intel-free-libs-and-python-apt-repo more effective ways to install MKL.

**Comment by rscohn2** [comment]
This command will install pytorch 1.2 with all Intel optimizations for CPU (with MKL-DNN statically linked in)

conda install pytorch cpuonly -c pytorch

**Comment by faroit** [comment]
> `conda install pytorch cpuonly -c pytorch`

@rscohn2 this solved it, thanks. Glad that my model still works flawlessly on 1.2
